### PR TITLE
[mxnet] [build, test] Release MXNet 1.7.0

### PR DESCRIPTION
**Pull request by ChaiBapchya**
*Issue #, if available:*

## Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Reviewer Checklist
* For reviewer, before merging, please cross-check:
- [ ] I've verified the code change on the config file mentioned above has already been reverted

*Description:*
Update the latest version of Apache MXNet [package renamed as aws-mx due to licensing concerns]

*Tests run:*

*DLC image/dockerfile:*

*Additional context:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+LABEL dlc_major_version="3"
+
+ARG PYTHON=python3
+ARG PIP=pip3
+ARG PYTHON_VERSION=3.6.10
+ARG MX_URL=https://gluon-pypi.s3-us-west-2.amazonaws.com/1.7.0/gluon-1.7.0-py2.py3-none-manylinux2014_x86_64.whl
+
+# The smdebug pipeline relies for following format to perform string replace and trigger DLC pipeline for validating
+# the nightly builds. Therefore, while updating the smdebug version, please ensure that the format is not disturbed.
+ARG SMDEBUG_VERSION=0.9.2
```
Comment (NihalHarish) [inline mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu:13]: The smdebug binary is not usually included in the inference container.

File: mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    gluoncv==0.6.0 \
+    urllib3==1.25.8 \
+    python-dateutil==2.8.0 \
+    tqdm==4.39.0 \
+    sagemaker-experiments==0.* \
+    PyYAML==5.3.1 \
+    mpi4py==3.0.2 \
+    "sagemaker-mxnet-training<5" \
+    ${MX_URL} \
+    smdebug==${SMDEBUG_VERSION} \
```
Comment (NihalHarish) [inline mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu:143]: Remove.

File: mxnet/inference/docker/1.7.0/py3/cu101/Dockerfile.gpu [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+LABEL dlc_major_version="3"
+
+ARG PYTHON=python3
+ARG PIP=pip3
+ARG PYTHON_VERSION=3.6.10
+ARG MX_URL=https://gluon-pypi.s3-us-west-2.amazonaws.com/1.7.0/gluon_cu101-1.7.0-py2.py3-none-manylinux2014_x86_64.whl
+
+# The smdebug pipeline relies for following format to perform string replace and trigger DLC pipeline for validating
+# the nightly builds. Therefore, while updating the smdebug version, please ensure that the format is not disturbed.
+ARG SMDEBUG_VERSION=0.9.2
```
Comment (NihalHarish) [inline mxnet/inference/docker/1.7.0/py3/cu101/Dockerfile.gpu:14]: Remove

File: mxnet/inference/docker/1.7.0/py3/cu101/Dockerfile.gpu [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    gluoncv==0.6.0 \
+    urllib3==1.25.8 \
+    python-dateutil==2.8.0 \
+    tqdm==4.39.0 \
+    sagemaker-experiments==0.* \
+    PyYAML==5.3.1 \
+    mpi4py==3.0.2 \
+    "sagemaker-mxnet-training<5" \
+    ${MX_URL} \
+    smdebug==${SMDEBUG_VERSION} \
```
Comment (NihalHarish) [inline mxnet/inference/docker/1.7.0/py3/cu101/Dockerfile.gpu:185]: Remove

File: mxnet/training/docker/1.7.0/py3/Dockerfile.cpu [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+LABEL dlc_major_version="3"
+
+ARG PYTHON=python3
+ARG PIP=pip3
+ARG PYTHON_VERSION=3.6.10
+ARG MX_URL=https://gluon-pypi.s3-us-west-2.amazonaws.com/1.7.0/gluon-1.7.0-py2.py3-none-manylinux2014_x86_64.whl
+
+# The smdebug pipeline relies for following format to perform string replace and trigger DLC pipeline for validating
+# the nightly builds. Therefore, while updating the smdebug version, please ensure that the format is not disturbed.
+ARG SMDEBUG_VERSION=0.9.2
```
Comment (NihalHarish) [inline mxnet/training/docker/1.7.0/py3/Dockerfile.cpu:13]: Bump the version to `0.9.3`

File: mxnet/training/docker/1.7.0/py3/cu101/Dockerfile.gpu [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+LABEL dlc_major_version="3"
+
+ARG PYTHON=python3
+ARG PIP=pip3
+ARG PYTHON_VERSION=3.6.10
+ARG MX_URL=https://gluon-pypi.s3-us-west-2.amazonaws.com/1.7.0/gluon_cu101-1.7.0-py2.py3-none-manylinux2014_x86_64.whl
+
+# The smdebug pipeline relies for following format to perform string replace and trigger DLC pipeline for validating
+# the nightly builds. Therefore, while updating the smdebug version, please ensure that the format is not disturbed.
+ARG SMDEBUG_VERSION=0.9.2
```
Comment (NihalHarish) [inline mxnet/training/docker/1.7.0/py3/cu101/Dockerfile.gpu:14]: Bump the version to `0.9.3`

**Review by NihalHarish — CHANGES_REQUESTED** [review]

**Comment by NihalHarish** [comment]
Is there going to also be a cu102 container?

File: mxnet/inference/docker/1.7.0/py3/cu101/Dockerfile.gpu (hunk 3 above)
Comment (ChaiBapchya) [reply mxnet/inference/docker/1.7.0/py3/cu101/Dockerfile.gpu:14]: Done

File: mxnet/inference/docker/1.7.0/py3/cu101/Dockerfile.gpu (hunk 4 above)
Comment (ChaiBapchya) [reply mxnet/inference/docker/1.7.0/py3/cu101/Dockerfile.gpu:185]: Done

File: mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu (hunk 2 above)
Comment (ChaiBapchya) [reply mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu:143]: Done

File: mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu (hunk 1 above)
Comment (ChaiBapchya) [reply mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu:13]: Done

File: mxnet/training/docker/1.7.0/py3/Dockerfile.cpu (hunk 5 above)
Comment (ChaiBapchya) [reply mxnet/training/docker/1.7.0/py3/Dockerfile.cpu:13]: Bumped!

File: mxnet/training/docker/1.7.0/py3/cu101/Dockerfile.gpu (hunk 6 above)
Comment (ChaiBapchya) [reply mxnet/training/docker/1.7.0/py3/cu101/Dockerfile.gpu:14]: Bumped!

**Comment by ChaiBapchya** [comment]
> Is there also going to be a cu102 container?

Not for the purpose of this PR. Probably we might add gluon-cu102 with 1.7.0 later if there's a compelling reason. Right now the aim is to get 1.7.0 pushed out.

**Comment by NihalHarish** [comment]
The tests executed by this PR do not test MxNet 1.7.

You'll need to change also update `mxnet/buildspec.yml`

File: mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu [hunk 7]
```diff
@@ -0,0 +1,129 @@
+FROM ubuntu:16.04
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="3"
```
Comment (arjkesh) [inline mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu:4]: This should be major version 1, as this is the first release of 1.7.0 containers

File: mxnet/inference/docker/1.7.0/py3/cu101/Dockerfile.gpu [hunk 8]
```diff
@@ -0,0 +1,126 @@
+FROM nvidia/cuda:10.1-cudnn7-runtime-ubuntu16.04
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="3"
```
Comment (arjkesh) [inline mxnet/inference/docker/1.7.0/py3/cu101/Dockerfile.gpu:4]: this should be major version 1

File: mxnet/training/docker/1.7.0/py3/Dockerfile.cpu [hunk 9]
```diff
@@ -0,0 +1,181 @@
+FROM ubuntu:16.04
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="3"
```
Comment (arjkesh) [inline mxnet/training/docker/1.7.0/py3/Dockerfile.cpu:4]: this should be major version 1

File: mxnet/training/docker/1.7.0/py3/cu101/Dockerfile.gpu [hunk 10]
```diff
@@ -0,0 +1,227 @@
+# Note: We need to install NCCL, cuDNN, and CUDA libraries since we are using base container.
+FROM nvidia/cuda:10.1-base-ubuntu16.04
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="3"
```
Comment (arjkesh) [inline mxnet/training/docker/1.7.0/py3/cu101/Dockerfile.gpu:5]: please change to major version 1

**Review by arjkesh — COMMENTED** [review]
Can you also add an example image? I believe this is missing and will be required for EKS tests.

**Review by classicsong — COMMENTED** [review]
The DGL related part is good to me

**Comment by arjkesh** [comment]
paging @classicsong to 

> The DGL related part is good to me

@classicsong we are installing 0.4.* - should we remove dgl from the container until mxnet 1.7.0 is explicitly supported?

File: mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu (hunk 7 above)
Comment (ChaiBapchya) [reply mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu]: good catch. updated

File: mxnet/inference/docker/1.7.0/py3/cu101/Dockerfile.gpu (hunk 8 above)
Comment (ChaiBapchya) [reply mxnet/inference/docker/1.7.0/py3/cu101/Dockerfile.gpu]: good catch. updated

File: mxnet/training/docker/1.7.0/py3/Dockerfile.cpu (hunk 9 above)
Comment (ChaiBapchya) [reply mxnet/training/docker/1.7.0/py3/Dockerfile.cpu]: updated

File: mxnet/training/docker/1.7.0/py3/cu101/Dockerfile.gpu (hunk 10 above)
Comment (ChaiBapchya) [reply mxnet/training/docker/1.7.0/py3/cu101/Dockerfile.gpu]: updated

**Comment by ChaiBapchya** [comment]
> Can you also add an example image? 

@arjkesh I'm sorry I didn't understand what's meant by example image. Could you give me an example? 
Within mxnet directory, i couldn't find an example image

File: mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+COPY mms-entrypoint.py /usr/local/bin/dockerd-entrypoint.py
+COPY config.properties /home/model-server
+
+RUN chmod +x /usr/local/bin/dockerd-entrypoint.py
+
+ADD https://raw.githubusercontent.com/aws/deep-learning-containers/master/src/deep_learning_container.py /usr/local/bin/deep_learning_container.py
+
+RUN chmod +x /usr/local/bin/deep_learning_container.py
+
+RUN curl https://aws-dlc-licenses.s3.amazonaws.com/aws-mxnet-1.6.0/license.txt -o /license.txt
```
Comment (rondogency) [inline mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu:125]: does this need to be updated also?

Comment (ChaiBapchya) [reply mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu:125]: @saimidu @arjkesh any idea?

**Comment by ChaiBapchya** [comment]
> > Can you also add an example image?
> 
> @arjkesh I'm sorry I didn't understand what's meant by example image. Could you give me an example?
> Within mxnet directory, i couldn't find an example image

Nevermind. I found that the example folder in training Dockerfiles. I added an example for v1.7.0 mx training py3 Dockerfile. @arjkesh

**Comment by ChaiBapchya** [comment]
```
Collecting gluonnlp==0.10.0
  Downloading gluonnlp-0.10.0.tar.gz (344 kB)
    ERROR: Command errored out with exit status 1:

     command: /usr/local/bin/python3.7 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-hum2oxjp/gluonnlp/setup.py'"'"'; __file__='"'"'/tmp/pip-install-hum2oxjp/gluonnlp/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-ph0d2xuo

         cwd: /tmp/pip-install-hum2oxjp/gluonnlp/

    Complete output (11 lines):

    Traceback (most recent call last):

      File "<string>", line 1, in <module>

      File "/usr/local/lib/python3.7/site-packages/setuptools/__init__.py", line 18, in <module>

        from setuptools.dist import Distribution
      
      File "/usr/local/lib/python3.7/site-packages/setuptools/dist.py", line 32, in <module>

        from setuptools import windows_support

      File "/usr/local/lib/python3.7/site-packages/setuptools/windows_support.py", line 2, in <module>

        import ctypes

      File "/usr/local/lib/python3.7/ctypes/__init__.py", line 7, in <module>

        from _ctypes import Union, Structure, Array

    ModuleNotFoundError: No module named '_ctypes'

    ----------------------------------------

ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

@leezu This seems like gluonnlp doesnt get installed using python37. Is that correct?

File: mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu (hunk 11 above)
Comment (ChaiBapchya) [reply mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu:125]: Yes @tejaschumbalkar is updating the license file I provided for 1.7.0.

**Comment by ChaiBapchya** [comment]
Testing if gluonnlp issue resolved with py36

File: mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu (hunk 11 above)
Comment (tejaschumbalkar) [reply mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu:125]: New license is available here: https://aws-dlc-licenses.s3.amazonaws.com/aws-mxnet-1.7.0/license.txt

**Comment by ChaiBapchya** [comment]
It went pass the gluonnlp install issue by switching to py36.
Now its failing to build horovod for MXNet.
Complains with
```
In file included from /usr/local/lib/python3.6/site-packages/mxnet/include/mkldnn/mkldnn.hpp:24:0,
                     from /usr/local/lib/python3.6/site-packages/mxnet/include/mxnet/ndarray.h:41,
                     from horovod/mxnet/mpi_ops.h:23,
                     from horovod/mxnet/mpi_ops.cc:20:
    /usr/local/lib/python3.6/site-packages/mxnet/include/mkldnn/dnnl.hpp:23:25: fatal error: dnnl_config.h: No such file or directory
    compilation terminated.
    error: command 'gcc' failed with exit status 1
```
I checked the mxnet binary does not contain the aforementioned file `dnnl_config.h` in `/mxnet/include/mkldnn` of the unzipped wheel. 
Looking into fixing the missing header file issue.

File: mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu (hunk 11 above)
Comment (ChaiBapchya) [reply mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu:125]: To make sure we comply with the new package naming, is it fine if you move the license to folder name `aws-mx-1.7.0`
We want to make sure we don't use "mxnet" for the package name or license.

**Comment by ChaiBapchya** [comment]
The issue was with a missing header file [`dnnl_config.h`] due to it not being placed in the right location `3rdparty/mkldnn/include/`. Fixed it while building mxnet binary & retriggered the pipeline.

File: mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu (hunk 11 above)
Comment (tejaschumbalkar) [reply mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu:125]: License moved to https://aws-dlc-licenses.s3.amazonaws.com/aws-mx-1.7.0/license.txt

**Comment by tejaschumbalkar** [comment]
Single node EKS test for inference are failing

**Review by NihalHarish — DISMISSED** [review]
The smdebug changes LGTM

File: mxnet/buildspec.yml [hunk 12]
```diff
@@ -54,28 +54,28 @@
...
-      build: &MXNET_GPU_TRAINING_PY2 false
-      image_size_baseline: 4899
-      device_type: &DEVICE_TYPE gpu
-      python_version: &DOCKER_PYTHON_VERSION py2
-      tag_python_version: &TAG_PYTHON_VERSION py27
-      tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *CUDA_VERSION, "-", *OS_VERSION ]
-      docker_file: !join [ docker/, *VERSION, /, *DOCKER_PYTHON_VERSION, /, *CUDA_VERSION, /Dockerfile., *DEVICE_TYPE ]
-      context:
-        <<: *TRAINING_CONTEXT
+    # BuildMXNetCPUTrainPy2DockerImage:
```
Comment (tejaschumbalkar) [inline mxnet/buildspec.yml:57]: Can we remove these if we are not building py2 images

File: mxnet/buildspec.yml [hunk 13]
```diff
@@ -98,41 +98,41 @@
...
-      base_image_name: BuildMXNetGPUTrainPy2DockerImage
-      device_type: &DEVICE_TYPE gpu
-      python_version: &DOCKER_PYTHON_VERSION py2
-      tag_python_version: &TAG_PYTHON_VERSION py27
-      tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *CUDA_VERSION, "-", *OS_VERSION,
-                   "-example" ]
-      docker_file: !join [ docker/, *VERSION, /, *DOCKER_PYTHON_VERSION, /example, /Dockerfile., *DEVICE_TYPE ]
-      context:
-        <<: *TRAINING_CONTEXT
+    # BuildMXNetCPUInferencePy2DockerImage:
```
Comment (tejaschumbalkar) [inline mxnet/buildspec.yml:101]: same here

**Comment by tejaschumbalkar** [comment]
Also, update the license url:  https://aws-dlc-licenses.s3.amazonaws.com/aws-mx-1.7.0/license.txt

File: mxnet/buildspec.yml (hunk 12 above)
Comment (ChaiBapchya) [reply mxnet/buildspec.yml]: removed!

File: mxnet/buildspec.yml (hunk 13 above)
Comment (ChaiBapchya) [reply mxnet/buildspec.yml]: removed!

**Comment by saimidu** [comment]
If there is going to be another iteration of this PR, please add the buildspec for 1.6.0 as a separate `buildspec-1-6.yml`. Please take a look at https://github.com/aws/deep-learning-containers/tree/master/pytorch for an example of how buildspecs for older framework versions are maintained.

If there is not going to be another iteration, please follow this PR up with another one to add that buildspec (this new PR will not trigger builds or tests.)

File: mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+# Specify accept-bind-to-port LABEL for inference pipelines to use SAGEMAKER_BIND_TO_PORT
+# https://docs.aws.amazon.com/sagemaker/latest/dg/inference-pipeline-real-time.html
+LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
+# Specify multi-models LABEL to indicate container is capable of loading and serving multiple models concurrently
+# https://docs.aws.amazon.com/sagemaker/latest/dg/build-multi-model-build-container.html
+LABEL com.amazonaws.sagemaker.capabilities.multi-models=true
```
Comment (saimidu) [inline mxnet/inference/docker/1.7.0/py3/Dockerfile.cpu:11]: Please verify that this is correct.

**Review by tejaschumbalkar — DISMISSED** [review]

**Review by saimidu — APPROVED** [review]

**Comment by tejaschumbalkar** [comment]
Seeing failures in SageMaker and EKS test builds

**Comment by ChaiBapchya** [comment]
@saimidu eks test failed again. timeout it seems again.
Can we disable this test and merge the PR? 
Also curious why do I see py3.8 in eks test logs?

**Comment by saimidu** [comment]
Do we know what the root cause of the EKS test failure is?

**Comment by ChaiBapchya** [comment]
The issue is with downloading one bert package from the s3-accelerate [localhost]. @tejaschumbalkar has more context on it and can update.

**Review by tejaschumbalkar — DISMISSED** [review]

**Review by tejaschumbalkar — APPROVED** [review]

**Pull request merged**